packages/liveblocks-core/src/lib/objectToQuery.ts

```diff
@@ -0,0 +1,222 @@
+/**
+ * Converts an object to a query string
+ * Example:
+ * ```ts
+ * const query = objectToQuery({
+      resolved: true,
+      metadata: {
+        status: "open",
+        priority: 3,
+        org: {
+          startsWith: "liveblocks:",
+        },
+      },
+});
+
+console.log(query);
+// resolved:true AND metadata["status"]:open AND metadata["priority"]:3 AND metadata["org"]^"liveblocks:"
+
+ * ```
+ *
+ *
+ */
+
+type FilterValue = string | number | boolean | { startsWith: string };
+
+type Filter = NumberFilter | StringFilter | BooleanFilter;
+
+type NumberFilter = {
+  key: string;
+  operator: ":";
+  value: number;
+};
+
+type StringFilter = {
+  key: string;
+  operator: ":" | "^";
+  value: string;
+};
+
+type BooleanFilter = {
+  key: string;
+  operator: ":";
+  value: boolean;
+};
+
+/**
+ * Converts an object to a query string
+ * @example
+ * ```ts
+ * const query = objectToQuery({
+ *  metadata: {
+ *    status: "open",
+ *    priority: 3,
+ *    org: {
+ *      startsWith: "liveblocks:",
+ *    },
+ *  },
+ * });
+ * console.log(query);
+ * // metadata["status"]:"open" AND metadata["priority"]:3 AND metadata["org"]^"liveblocks:"
+ * ```
+ * @example
+ * ```ts
+ * const query = objectToQuery({
+ *  resolved: true,
+ *  roomId: {
+ *   startsWith: "engineering:",
+ *  },
+ * });
+ * console.log(query);
+ * // resolved:true AND roomId^"engineering:"
+ * ```
+ *
+ */
+export function objectToQuery(obj: {
+  [key: string]: FilterValue | { [key: string]: FilterValue };
+}): string {
+  let filterList: Filter[] = [];
+  const entries = Object.entries(obj);
+
+  /**
+   * Case 1: Simple key value pairs (equal operator)
+   * @example
+   * ```ts
+   * { key: "foo" }
+   * ```
+   */
+  const keyValuePairs = entries.filter(([_, value]) =>
+    isSimpleValue(value)
+  ) as [string, string | number | boolean][];
```

Comment:
You can avoid doing this type-cast here…

```suggestion
  );
```